{% extends "base.html" %}
{% block body %}

<div class="container">
    <h3>Overview</h3>

    <p>The ExPheWas browser showcases the results of a phenome-wide association study (PheWAS) using an approach that models the joint effect of variants at protein coding genes. In the current data release, we tested the association between 26,616 protein coding or lincRNA genes and 1,746 phenotypes available in the UK Biobank.</p>

    <p>Briefly, we used a principal component analysis (PCA) of common variant genotypes to reduce the effective number of variables needed to capture statistical variance in the genotypes. Association testing is then done by assessing the improvement in goodness of fit from including the gene based principal components to a regression model. The experimental setup is described in more details in the figure below:</p>

    <p style="text-align: center;"><img style="max-width: 100%; height: auto" src="{{ url_for('backend_blueprint.static', filename='v1_analysis_summary_web.png') }}" /></p>

    <p>PheWAS results can then be browsed <a href="{{ url_for("backend_blueprint.get_genes") }}">by gene</a> or <a href="{{ url_for("backend_blueprint.get_outcomes") }}">by phenotype</a>.</p>

    <div class="row" style="margin-bottom: 30px;">
      <div class="col-sm-4">
        <div class="card">
          <div class="card-body">
            <h5 class="card-title">Phenotype page</h5>
            <p class="card-text">See all genes associated with a given phenotype.</p>
            <a href="{{ url_for('backend_blueprint.get_outcomes') }}" class="btn btn-primary">List of phenotypes</a><br/>
            <a href="{{ url_for('backend_blueprint.get_random_outcome') }}" class="card-link">(or random example)</a>
          </div>
        </div>
      </div>
      <div class="col-sm-4">
        <div class="card">
          <div class="card-body">
            <h5 class="card-title">Gene page</h5>
            <p class="card-text">See all phenotypes associated with a given gene.</p>
            <a href="{{ url_for('backend_blueprint.get_genes') }}" class="btn btn-primary">List of genes</a><br />
            <a href="{{ url_for('backend_blueprint.get_random_gene') }}" class="card-link">(or random example)</a>
          </div>
        </div>
      </div>
      <div class="col-sm-4">
        <div class="card">
          <div class="card-body">
            <h5 class="card-title"><i>cis</i>-Mendelian randomization</h5>
            <p class="card-text">Estimate the causal (under the instrumental variable assumptions) effect of an exposure on an outcome based only on a single gene.</p>
            <a href="{{ url_for('backend_blueprint.cis_mr') }}" class="btn btn-primary">cis-MR</a>
          </div>
        </div>
      </div>
    </div>

    <h3>Citation</h3>
    <p>If you use ExPheWas in your research, please consider citing:</p>
    <div class="citation">
      <p><b><i>ExPheWas</i>: a browser for gene-based pheWAS associations.</b><br /> Legault, M.-A., Lemieux Perreault, L.-P., & Dub√©, M.-P. (2021). <br />Preprint on <i>medRxiv</i>. <a href="https://doi.org/10.1101/2021.03.17.21253824">https://doi.org/10.1101/2021.03.17.21253824</a></p>
    </div>

    <h3>Data release information</h3>
    <p>We use distinct version numbers for the data release and the browser. This enables us to improve on the browser while unambiguously referring to the underlying data. You may refer to the data release version when reporting ExPheWas results and we are commited to maintaining stable access to previous data releases. There are currently two data releases (<a href="https://exphewas.ca/v0">v0</a> and <a href="https://exphewas.ca/v1">v1</a>).</p>

    <p>This browser uses data release <span class="badge bg-success text-light">v{{ meta.db.version }}</span> ({{ meta.db.date }})</p>
    <p>{{ meta.db.comments|safe }}</p>

    <h3>Support</h3>

    <p>For technical assistance or feature requests, we recommend you open a ticket on the project's <a href="https://github.com/pgxcentre/ExPheWAS/issues">Github page</a>. For other inquiries, you may contact the corresponding author of the manuscript.</p>

</div>

{% endblock %}

{% block other_script %}
{% endblock %}
